google/cloud/pubsub/subscriber.h

```diff
@@ -123,10 +122,44 @@ class Subscriber {
    * [std-function-link]:
    * https://en.cppreference.com/w/cpp/utility/functional/function
    */
-  future<Status> Subscribe(std::function<void(Message, AckHandler)> cb) {
+  future<Status> Subscribe(ApplicationCallback cb) {
     return connection_->Subscribe({std::move(cb)});
   }
 
+  /**
+   * Creates a new session to receive messages from @p subscription using
+   * exactly-once delivery.
+   *
+   * @par Idempotency
+   * @parblock
+   * This is an idempotent operation; it only reads messages from the service.
+   * Will make multiple attempts to start a connection to the service, subject
+   * to the retry policies configured in the `SubscriberConnection`. Once a
+   * successful connection is established the library will try to resume the
+   * connection even if the connection fails with a permanent error. Resuming
+   * the connection is subject to the retry policies as described earlier.
+   *
+   * Note that calling `ExactlyOnceAckHandler::ack()` and/or
+   * `ExactlyOnceAckHandler::nack()` have their own rules with respect to
+   * retrying. Check the documentation of these functions for details.
+   * @endparblock
+   *
+   * @par Example
+   * @snippet samples.cc exactly-once-subscribe
+   *
+   * @param cb the callable invoked when messages are received.
+   * @return a future that is satisfied when the session will no longer receive
+   *     messages. For example, because there was an unrecoverable error trying
+   *     to receive data. Calling `.cancel()` in this object will (eventually)
+   *     terminate the session and satisfy the future.
+   *
+   * [std-function-link]:
+   * https://en.cppreference.com/w/cpp/utility/functional/function
+   */
+  future<Status> Subscribe(ExactlyOnceApplicationCallback cb) {
```

Comment:
should this be called `ExactlyOnceSubscribe` to match the connection function?

Or do you not want conflicting names in the connection layer, because it makes writing the mock `EXPECT_CALL`s kind of annoying?

Reply:
The second.  I want the public API to the easy to use (call `Subscribe()` and the overloads will figure it out), and the mocks to also be easy to use. Add some comments to explain what is going on.

Reply:
Makes sense.